### PR TITLE
perf(colocate): add bidirectional GPU weight handoff

### LINKED ISSUE #169: [Task 24] Reduce Colocate Phase-Switch Overhead with Bidirectional GPU Weight Handoff
## Background and Motivation

In the four-GPU Qwen3-VL-4B colocate baseline, rollout and training share the same GPUs. Therefore, phase-switch overhead directly blocks end-to-end training.

| Baseline metric | Time | Percentage of step time |
|---|---:|---:|
| Mean step time | 22.361 s | 100% |
| Rollout -> Train | 2.736 s | 12.2% |
| Train -> Rollout | 10.530 s | 47.1% |
| Total transition proxy | 13.266 s | 59.3% |

The legacy transition proxy is `sleep + wake_up + update_weights`. These values are host-observed wall times, not GPU kernel-active times.

Approximately 59% of the baseline step time is spent switching phases. The existing path also stages actor weights through Host memory and repeatedly releases and restores phase-specific states.

This task proposes bidirectional GPU weight handoff, selective non-weight memory release, and overlap with optimizer offload to reduce this blocking overhead.

## Repository and Development Branch

- Repository: `redai-infra/Relax`
- Upstream base commit: `cfda05351d4b137f60e2b71d048beab1316d9d9f`
- Fork: `fxfxfxfxfxfxfxfx/Relax`
- Development branch: `task24-colocate-weight-handoff`
- Current development commit: `6e55ad0f2fd854a55d015fd06e12d76fec3dc346`
- The A/B benchmark uses the same implementation commit and differs only in `--colocate-weight-handoff`.

## Environment

- Base image: `lmsysorg/sglang:v0.5.12.post1-cu129`
- OS: Ubuntu 22.04.5
- Python: 3.12.3
- CUDA: 12.9
- NVIDIA driver: 580.105.08
- PyTorch: 2.11.0+cu129
- Ray: 2.56.1
- Transformers: 5.6.0
- SGLang: 0.5.12.post1
- SGLang base commit: `5a15cde858ea09b77116212a39356f2fc51b8584`
- Megatron-Bridge commit: `2faedbf6fe3c422835a44b2b360cadcb2a116a54`
- Megatron-LM commit: `85bced0ae6ab46f61a0fd774074a3273daf6ae02`
- GPU: 4 x NVIDIA A800 80GB PCIe
- CPU: Intel Xeon Gold 6348, 112 logical CPUs
- System memory: 1 TiB, no swap

## Entry Script and Commands

Entry script:

```text
scripts/training/multimodal/run-qwen3-vl-4B-8xgpu.sh
```

### Common Configuration

```bash
export CUDA_VISIBLE_DEVICES=0,1,2,3
export NUM_GPUS=4
export TASK_GPUS=4

### PR DESCRIPTION
**Pull request by fxfxfxfxfxfxfxfx**
## What

- Closes #169.
- Adds an opt-in `--colocate-weight-handoff` path for Qwen3-VL-4B colocate training.
- Reuses live GPU weights in both phase transitions instead of staging actor weights through Host memory.
- Releases phase-specific non-weight state and overlaps optimizer offload with the train-to-rollout handoff.
- Keeps the existing colocate path unchanged when the switch is disabled.

## Why

On the four-GPU colocate baseline, rollout and training share the same GPUs and phase transitions block the full training step. The legacy `sleep + wake_up + update_weights` transition proxy takes 13.266 seconds out of a 22.361-second mean step, or 59.3% of the host-observed step wall time. Train-to-rollout is the dominant transition at 10.530 seconds.

The existing path also stages actor weights through Host memory and restores more engine state than each destination phase needs. This PR introduces direct GPU-to-GPU layout conversion and selective state release to reduce that serialized overhead.

## How

- Adds `ColocateWeightHandoff` to coordinate rollout-to-train and train-to-rollout transitions, memory preflight checks, source lifetime, failure cleanup, and timing metrics.
- Splits rollout-to-train into `prepare_train()` and `activate_train_weights()` so reference work and non-weight release can overlap before SGLang weights are imported into Megatron.
- Adds `UpdateWeightFromSGLang` and extends the Megatron Bridge conversion path to write the current TP2 shards directly from exported SGLang CUDA tensors.
- Adds a Qwen3-VL inverse mapping in the versioned SGLang patch for text QKV, gate/up projections, vision weights, and tied embeddings.
- Exposes read-only same-node CUDA IPC views through `export_weights_to_tensor()` without an entire-model staging buffer.
- Reuses Megatron's optimizer state offloader through a lazy `enable_state_offload()` entry point.
- Passes `weights`, `kv_cache`, and `cuda_graph` release tags through the Relax SGLang wrapper.
- Keeps the feature disabled by default and rejects unsupported layouts before target allocation.

### Supported V1 configuration

- Qwen3-VL-4B BF16 in colocate mode
- Megatron TP2 / PP1 / EP1
- SGLang TP1
- No quantization, speculative decoding, critic, teacher, or `keep-old-actor`

Disaggregated execution, cross-node reverse handoff, generic converter registration, and long-lived shared parameter addresses are out of scope.

### Environment

- Base: `cfda05351d4b137f60e2b71d048beab1316d9d9f`
- Hardware: 4 x NVIDIA A800 80GB PCIe
- Python 3.12.3, CUDA 12.9, PyTorch 2.11.0+cu129
- SGLang 0.5.12.post1, Megatron-Bridge `2faedbf`, Megatron-LM `85bced0`
- Model: Qwen3-VL-4B-Instruct BF16
- Layout: Megatron TP2/PP1/EP1 and four SGLang TP1 engines

### Reproduction

```bash
export CUDA_VISIBLE_DEVICES=0,1,2,3
export NUM_GPUS=4 TASK_GPUS=4
export MODEL_DIR=/root/autodl-tmp/relax-workspace/models
export DATA_DIR=/root/autodl-tmp/relax-workspace/datasets
export NUM_ROLLOUT=23 ROLLOUT_BATCH_SIZE=8
export N_SAMPLES_PER_PROMPT=1 GLOBAL_BATCH_SIZE=8
export ROLLOUT_MAX_PROMPT_LEN=1024 ROLLOUT_MAX_RESPONSE_LEN=256
export MAX_TOKENS_PER_GPU=8192 LOG_PROBS_MAX_TOKENS_PER_GPU=8192
export OMP_NUM_THREADS=24 MKL_NUM_THREADS=24 OPENBLAS_NUM_THREADS=24

ENABLE_COLOCATE_WEIGHT_HANDOFF=0 \
bash scripts/training/multimodal/run-qwen3-vl-4B-8xgpu.sh

ENABLE_COLOCATE_WEIGHT_HANDOFF=1 \
bash scripts/training/multimodal/run-qwen3-vl-4B-8xgpu.sh
```

Both sides use the same model, data, hardware, batch sizes, sequence limits, sampling settings, three warmup steps, and 20 measured steps. Startup, checkpoint loading, CUDA graph capture, and dataset scanning are excluded.

### Test results

- `python -m pytest -q tests/backends/megatron/test_colocate_weight_handoff.py tests/utils/test_arguments_opd_teacher_colocate.py`: 19 passed.
- `python -m pytest -q test/registered/unit/models/test_qwen3_vl_weight_export.py`: 3 passed.
- Targeted `ruff check`: passed.
- `bash -n scripts/training/multimodal/run-qwen3-vl-4B-8xgpu.sh`: passed.
- Versioned SGLang patch applies cleanly to `5a15cde`.
- Versioned Megatron patch applies cleanly to the merged Bridge/MCore source tree.

### Performance

Weighted throughput is total actor-train tokens divided by total measured step wall time.

| Metric | Switch off | Switch on | Change |
|---|---:|---:|---:|
| Mean step time | 22.361 s | 12.645 s | -43.45% |
| P50 step time | 22.505 s | 12.939 s | -42.51% |
| P95 step time | 22.899 s | 13.271 s | -42.05% |
| Weighted total-token throughput | 193.948 tok/s | 343.536 tok/s | +77.13% |
| Response-token throughput | 80.316 tok/s | 142.548 tok/s | +77.48% |
| Actor train MFU | 5.939% | 6.473% | +0.533 pp |
| Rollout-to-train | 2.736 s | 1.673 s | -38.84% |
| Train-to-rollout | 10.530 s | 2.822 s | -73.21% |
| Total transition proxy | 13.266 s | 4.495 s | -66.12% |

Optimized transition components:

| Component | Mean time |
|---|---:|
| SGLang non-weight offload | 0.153 s |
| SGLang-to-Megatron conversion | 0.531 s |
| Megatron-to-SGLang conversion | 0.740 s |
| Optimizer D2H offload | 1.786 s |
| Train weight release | 0.639 s |
| Rollout KV/graph resume | 0.315 s |

The optimized train-to-rollout components sum to 3.479 seconds while the enclosing transition is 2.822 seconds, indicating 0.658 seconds of measured overlap. Weight handoff transferred 0.000 GiB through Host staging. Optimized peak GPU memory was 49.415 GiB mean, 49.935 GiB P95, and 49.981 GiB maximum.

The phase and transition values above are host-observed wall times. The wall-time split changed from 85.51% wait-side / 14.49% train to 58.06% / 41.94%.

### Correctness and cleanup

- Mean train/rollout log-prob absolute difference: 0.01777 off vs. 0.01719 on.
- Mean mismatch KL: 0.001274 off vs. 0.000922 on.
- A separate `n=2` validation completed with nonzero steady-state gradient norms of 1.273 and 2.041.
- Every optimized transition exported 605 live SGLang tensors and updated 713 Megatron parameters.
- All jobs exited without CUDA OOM, NaN/Inf, Ray task failure, or actor failure.
- No Ray, SGLang, Megatron, or GPU compute process remained after shutdown.

- [ ] `pre-commit run --all-files` passes
- [x] Targeted tests pass
- [x] New tests added
- [x] Entry script and default-off configuration updated

## Risk and Rollback

- The optimized transition temporarily requires both source and destination weight layouts to fit on GPU. A preflight check fails before target allocation when the configured memory margin is insufficient.
- Source weights are retained until target validation succeeds. Failed handoffs release incomplete target state and keep the source phase usable.
- The implementation is intentionally restricted to the Qwen3-VL-4B TP2/TP1 layout in V1.
- Rollback requires only removing `--colocate-weight-handoff` or leaving `ENABLE_COLOCATE_WEIGHT_HANDOFF=0`, which restores the existing `sleep/wake/update_weights` path.

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update
- [ ] Refactoring
- [x] Performance improvement
- [ ] CI/CD or build changes

## Screenshots / Logs

Before/after curves and sanitized benchmark logs will be attached to this Draft PR.

### CODE REVIEW

**Comment by xiaoliang0601** [comment]
请补充正确性验证（step 1 ~ 50 的 loss、grad norm、reward、mismatch 的对比曲线）。

**Comment by fxfxfxfxfxfxfxfx** [comment]
### Correctness validation: steps 1-50

Added the requested comparison curves for policy loss, gradient norm, raw reward, and train/rollout mismatch KL.

Both runs used the same Relax commit (`64cff35`), Qwen3-VL-4B BF16 model, dataset, 4 × A800 80GB GPUs, Megatron TP2, SGLang TP1, batch/sequence settings, seeds, and measurement interval. The only functional difference was whether `--colocate-weight-handoff` was enabled.
image

| Metric | Handoff OFF, mean ± std | Handoff ON, mean ± std |
|---|---:|---:|
| Policy loss | 0.002239 ± 0.011718 | -0.000231 ± 0.013321 |
| Gradient norm | 1.3162 ± 1.6069 | 0.8593 ± 1.2557 |
| Raw reward | 0.3050 ± 0.1991 | 0.2850 ± 0.2329 |
| Mismatch KL | 0.001263 ± 0.001261 | 0.001064 ± 0.001097 |

Rollout sampling is stochastic, so pointwise identity is not expected. The 95% confidence intervals of the means overlap for all four metrics. Policy loss remains centered near zero, reward stays in the same range, and mismatch KL remains at the same `1e-3` scale.

Zero loss/gradient steps occur when both samples in a GRPO prompt group receive the same reward, resulting in zero groupwise advantage.

Both runs completed steps 0-50 successfully. All plotted values are finite, the enabled run completed 51 consecutive bidirectional handoff cycles, and no OOM or residual GPU/Ray process was observed.
